pkg/tasks/metricsserver.go

```diff
@@ -0,0 +1,210 @@
+		}
+	}
+	{
+		dep, err := t.factory.MetricsServerDeployment()
+		if err != nil {
+			return errors.Wrap(err, "initializing MetricsServer Deployment failed")
+		}
+
+		err = t.client.CreateOrUpdateDeployment(ctx, dep)
+		if err != nil {
+			return errors.Wrap(err, "reconciling MetricsServer Deployment failed")
+		}
+	}
+	{
+		sm, err := t.factory.MetricsServerServiceMonitor()
+		if err != nil {
+			return errors.Wrap(err, "initializing MetricsServer ServiceMonitors failed")
+		}
+
+		err = t.client.CreateOrUpdateServiceMonitor(ctx, sm)
+		if err != nil {
+			return errors.Wrapf(err, "reconciling %s/%s ServiceMonitor failed", sm.Namespace, sm.Name)
+		}
+	}
+	{
+		pdb, err := t.factory.MetricsServerPodDisruptionBudget()
+		if err != nil {
+			return errors.Wrap(err, "initializing MetricsServer PodDisruptionBudget failed")
+		}
+
+		if pdb != nil {
+			err = t.client.CreateOrUpdatePodDisruptionBudget(ctx, pdb)
+			if err != nil {
+				return errors.Wrap(err, "reconciling MetricsServer PodDisruptionBudget failed")
+			}
+		}
+	}
+	{
+		api, err := t.factory.MetricsServerAPIService()
+		if err != nil {
+			return errors.Wrap(err, "initializing MetricsServer APIService failed")
+		}
+
+		err = t.client.CreateOrUpdateAPIService(ctx, api)
+		if err != nil {
+			return errors.Wrap(err, "reconciling MetricsServer APIService failed")
+		}
+	}
+
+	return t.removePrometheusAdapterResources(ctx)
+}
+
+func (t *MetricsServerTask) removePrometheusAdapterResources(ctx context.Context) error {
+	pa := NewPrometheusAdapterTask(ctx, t.namespace, t.client, false, t.factory, t.config)
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "prometheus-adapter",
+			Namespace: t.namespace,
+		},
+	}
+
+	{
+		pdb, err := pa.factory.PrometheusAdapterPodDisruptionBudget()
+		if err != nil {
+			return errors.Wrap(err, "initializing PrometheusAdapter PodDisruptionBudget failed")
+		}
+
+		if pdb != nil {
+			err = pa.client.DeletePodDisruptionBudget(ctx, pdb)
+			if err != nil {
+				return errors.Wrap(err, "deleting PrometheusAdapter PodDisruptionBudget failed")
+			}
+		}
+	}
+	{
+		sm, err := pa.factory.PrometheusAdapterServiceMonitor()
+		if err != nil {
+			return errors.Wrap(err, "initializing PrometheusAdapter ServiceMonitors failed")
+		}
+
+		err = pa.client.DeleteServiceMonitor(ctx, sm)
+		if err != nil {
+			return errors.Wrapf(err, "deleting %s/%s ServiceMonitor failed", sm.Namespace, sm.Name)
+		}
+	}
+	{
+		s, err := pa.factory.PrometheusAdapterService()
+		if err != nil {
+			return errors.Wrap(err, "initializing PrometheusAdapter Service failed")
+		}
+
+		err = pa.client.DeleteService(ctx, s)
+		if err != nil {
+			return errors.Wrap(err, "deleting PrometheusAdapter Service failed")
+		}
+	}
+	{
+		err := pa.client.DeleteDeployment(ctx, d)
+		if err != nil {
+			return errors.Wrap(err, "deleting PrometheusAdapter Deployment failed")
+		}
+	}
+
+	// TODO(slashpai): Add steps to remove other resources
```

Comment:
Is this `TODO` still valid? I was wondering about the p-a `APIService`. I thought there can only be on registered `APIService` for a given `group`. I might be missing details there though.

Reply:
I put that as TODO to do recheck :). I think we will have some more roles to remove

Reply:
API Service we are updating in metrics server task create step. Cannot delete that one since same name